homeassistant/components/image_processing/__init__.py

```diff
@@ -0,0 +1,253 @@
+CONF_CONFIDENCE = 'confidence'
+
+DEFAULT_TIMEOUT = 10
+DEFAULT_CONFIDENCE = 80
+
+SOURCE_SCHEMA = vol.Schema({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_SOURCE): vol.All(cv.ensure_list, [SOURCE_SCHEMA]),
+    vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
+    vol.Optional(CONF_CONFIDENCE, default=DEFAULT_CONFIDENCE):
+        vol.All(vol.Coerce(float), vol.Range(min=0, max=100))
+})
+
+SERVICE_SCAN_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+
+def scan(hass, entity_id=None):
+    """Force process a image."""
+    data = {ATTR_ENTITY_ID: entity_id} if entity_id else None
+    hass.services.call(DOMAIN, SERVICE_SCAN, data)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup image processing."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass, SCAN_INTERVAL)
+
+    yield from component.async_setup(config)
+
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file,
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    @asyncio.coroutine
+    def async_scan_service(service):
+        """Service handler for scan."""
+        image_entities = component.async_extract_from_service(service)
+
+        update_task = [entity.async_update_ha_state(True) for
+                       entity in image_entities]
+        if update_task:
+            yield from asyncio.wait(update_task, loop=hass.loop)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SCAN, async_scan_service,
+        descriptions.get(SERVICE_SCAN), schema=SERVICE_SCAN_SCHEMA)
+
+    return True
+
+
+class ImageProcessingEntity(Entity):
+    """Base entity class for image processing."""
+
+    timeout = DEFAULT_TIMEOUT  # timeout for requests
+
+    @property
+    def camera_entity(self):
+        """Return camera entity id from process pictures."""
+        return None
+
+    @property
+    def processing_class(self):
+        """Return the class of this entity from PROCESSING_CLASSES."""
+        return None
+
+    def process_image(self, image):
+        """Process image."""
+        raise NotImplementedError()
+
+    def async_process_image(self, image):
+        """Process image.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.loop.run_in_executor(None, self.process_image, image)
+
+    @property
+    def state_attributes(self):
+        """Return device specific state attributes."""
+        attr = {}
+
+        if self.processing_class is not None:
+            attr['processing_class'] = self.processing_class
+
+        return attr
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update image and process it.
+
+        This method is a coroutine.
+        """
+        websession = async_get_clientsession(self.hass)
```

Comment:
Can we extract fetching the image from an entity to a standalone method that lives in the camera class?

```python
def async_get_image(hass, entity_id):
    …
```